### DISCUSSION TITLE
Separating Code from Data in the Schema

### DISCUSSION THREAD

**Discussion by mobsense** [sensedeep/dynamodb-onetable#173] [Ideas]
# Separating Code Data in the Schema

This is a discussion starter about improving the separation of the schema and code logic so that the schema can be declarative and not contain embedded code. This should be achieved while still providing centralized control over validation, value templates, transformations and providing for custom needs.

Some of these embedded functions have been marked to be DEPRECATED, but we've not had a consistent or comprehensive solution to replace them adequately. This is the proposal to do that.

## Why?

We want the Schema to be 100% declarative and not contain embedded code for 2 reasons:

1. It permits the schema to be persisted in JSON and thus be created and managed by third party design tools. This is going to be increasingly important as several companies are now creating such tools. We want to be able to persist the schema in the table itself so the table becomes self-describing. This is essential for table migrations.
2. It is also good practice to keep code, as code, and the schema declarative. Mixing concerns is rarely good.

## Current Status

The schema currently supports three types of embedded functions. Albeit in deprecated form with warnings.

The embedded functions are:

* Transform functions
* Validation functions
* Value template functions

Transform functions are being used to store data values in custom formats in the table. Validation functions are being used to perform custom validations and multi-field validations across a property set. Value template functions are being employed by users to compute complex, overridden sort key values. 

## Solution

The proposed solution is to provide these functions as callbacks defined via the Table constructor.  This will centralize them, without embedding code in the schema.

For example

```javascript
const table = new Table({
    client: client,
    name: 'MyTable',
    schema: MySchema,
    transform: (model, operation, item, properties, params, raw) => {
    },
    validate: (model, properties, params) => {
    },
    value: (model, fieldName, properties, params) => {
    }
})
```

These callbacks can be inlined in the constructor as shown above when you need just a few tweaks to the default behavior. You can of source externalize them into separate functions and if you are subclassing Models, you can switch on the model.name and invoke relevant code in your Model classes.

NOTE: for most cases, the default OneTable behavior should suffice. These functions are for custom extensions to the default behavior.

## Transform

The optional Table `transform` function will be invoked on read and write requests to transform data before reading or writing to the table. The transform function can be used for custom storage formats or to assist with data migrations. The transform function can modify the item as it sees fit and return the modified item. The transform function is invoked in addition the the internal transformers. The invocation signature is:

```javascript
item = transform(model, operation, item, properties, params, raw)
```

Where `operation` is set to 'read' or 'write'. The `params` and `properties` are the original params and properties provided to the API call. When writing, the `item` will contain the already transformed properties by the internal transformers. You can overwrite the value in `item` using your own custom transformation logic using property values from `properties`.

When reading, the `item` will contain the already transformed read data and the `raw` parameter will contain the raw data as read from the table before conversion into Javascript properties in `item` via the internal transformers.

You can also use a `params.transform` with many Model APIs. See [Model API Params](#model-api-params) for details.

## Validate

The optional Table `validate` function will be invoked on requests to enable property validation before writing to the table.
The invocation signature is:

```javascript
details = validate(model, properties, params)
```

The validation function must return a map of validation messages for properties that fail validation checks. The map is indexed by the property field name. If the validate function returns a non-empty map, the validation is deemed to have failed and a `OneError` exception will be thrown that includes the details.

```
new OneError(`Validation Error for "${this.name}"`, {validation: details, code: 'Validation'})
```

## Value

The optional Table `value` function will be invoked for fields that define a value template in the schema and set the template to the value: `true`.

The value template can be used to compute the value of an attribute based on the property set and other app data.

```javascript
attributeValue = value(model, fieldName, properties, params)
```

**Comment by rjmackay** [comment]
What about including the field type in the `transform` function? That would allow for custom types based on the schema without having to detect the value type (in my case I'd use that for storing Luxon DateTime's)

**Reply by rjmackay** [reply]
Although I guess that doesn't quite work since `transform` is applied to an entire operation. It would be useful to have something like a field level transform. ie. applied on both read and write, but per field similar to `value`

**Reply by mobsense** [reply]
The reason transform is for the entire item is you may need other property values to combine for a field value.

You also have the model.name which indicates the type.

**Reply by rjmackay** [reply]
I'm was referring to value type (ie. string/boolean/date) not model type. I could inspecting the model schema to find all fields of `type: CustomType` or inspecting the raw value, but it's be nice to have the option of both tranforming on the entire item or individual fields depending on the scope.

**Reply by mobsense** [reply]
We could provide a model helper to fish that out. e.g.

    model.getType('field')

I think that would need to return

    model.block.fields[fieldName].type

Perhaps it would be useful to have getField too that would return the field structure with type, name, and other settings.

When your transform is invoked, you can elect to modify just the fields you want, or none and just return what you were given.

**Reply by rjmackay** [reply]
A `getField` helper makes sense to me
